sycl/source/detail/scheduler/commands.cpp

```diff
@@ -2106,7 +2106,12 @@ pi_int32 enqueueImpKernel(
   auto ContextImpl = Queue->getContextImplPtr();
   auto DeviceImpl = Queue->getDeviceImplPtr();
   RT::PiKernel Kernel = nullptr;
-  std::mutex *KernelMutex = nullptr;
+  // Cacheable kernels use per-kernel mutexes that will be fetched from the
+  // cache, others (e.g. interoperability kernels) share a single mutex.
+  // TODO consider adding a PiKernel -> mutex map for allowing to enqueue
+  // different PiKernel's in parallel.
+  static std::mutex NoncacheableEnqueueMutex;
```

Comment:
It would be probably better to have an `std::mutex`stored in `kernel_impl`. So, we do not serialize submission of different `sycl::kernel`s.

Reply:
@romanovvlad I tend to agree, although wouldn't this mutex need to be tied to `pi_kernel` like described in the comment above? Otherwise we will run into issues if two interop `sycl::kernel` are created with the same native kernel.

Reply:
Sounds reasonable.